diff --git a/1.3/indra/newview/llvoavatar.cpp b/1.3/indra/newview/llvoavatar.cpp
index 2036339..159ca1c 100644
--- a/1.3/indra/newview/llvoavatar.cpp
+++ b/1.3/indra/newview/llvoavatar.cpp
@@ -4956,7 +4956,7 @@ LLVector3 LLVOAvatar::getVolumePos(S32 joint_index, LLVector3& volume_offset)
 //-----------------------------------------------------------------------------
 LLJoint* LLVOAvatar::findCollisionVolume(U32 volume_id)
 {
-	if ((S32)volume_id > mNumCollisionVolumes)
+	if ((S32)volume_id > mNumCollisionVolumes || (S32)volume_id<0)
 	{
 		return NULL;
 	}
