diff --git a/trunk/indra/llplugin/CMakeLists.txt b/trunk/indra/llplugin/CMakeLists.txt
index d7f28a8..a27658d 100644
--- a/trunk/indra/llplugin/CMakeLists.txt
+++ b/trunk/indra/llplugin/CMakeLists.txt
@@ -64,4 +64,8 @@ list(APPEND llplugin_SOURCE_FILES ${llplugin_HEADER_FILES})
 
 add_library (llplugin ${llplugin_SOURCE_FILES})
 
+add_dependencies(llplugin
+  prepare
+)
+
 add_subdirectory(slplugin)
diff --git a/trunk/indra/media_plugins/base/CMakeLists.txt b/trunk/indra/media_plugins/base/CMakeLists.txt
index e4230be..6aa8be3 100644
--- a/trunk/indra/media_plugins/base/CMakeLists.txt
+++ b/trunk/indra/media_plugins/base/CMakeLists.txt
@@ -51,3 +51,7 @@ add_library(media_plugin_base
     ${media_plugin_base_SOURCE_FILES}
 )
 
+add_dependencies(media_plugin_base
+  prepare
+)
+
