Index: snowglobe-1.4/indra/media_plugins/gstreamer010/llmediaimplgstreamervidplug.cpp
===================================================================
--- snowglobe-1.4.orig/indra/media_plugins/gstreamer010/llmediaimplgstreamervidplug.cpp	2010-05-10 11:34:01.000000000 +0000
+++ snowglobe-1.4/indra/media_plugins/gstreamer010/llmediaimplgstreamervidplug.cpp	2010-05-10 11:46:13.000000000 +0000
@@ -512,11 +512,10 @@
    some g++ versions buggily avoid __attribute__((constructor)) functions -
    so we provide an explicit plugin init function.
  */
-void gst_slvideo_init_class (void)
-{
+
 #define PACKAGE "packagehack"
 	// this macro quietly refers to PACKAGE internally
-	static GST_PLUGIN_DEFINE (GST_VERSION_MAJOR,
+	       GST_PLUGIN_DEFINE (GST_VERSION_MAJOR,
 				  GST_VERSION_MINOR,
 				  "private-slvideoplugin", 
 				  "SL Video sink plugin",
@@ -524,6 +523,11 @@
 				  "Second Life",
 				  "http://www.secondlife.com/");
 #undef PACKAGE
+
+void gst_slvideo_init_class (void)
+{
+
+
 	ll_gst_plugin_register_static (&gst_plugin_desc);
 	DEBUGMSG("CLASS INIT");
 }
