diff --git a/trunk/indra/cmake/FindJsonCpp.cmake b/trunk/indra/cmake/FindJsonCpp.cmake
index 9d16f2a..205aa33 100644
--- a/trunk/indra/cmake/FindJsonCpp.cmake
+++ b/trunk/indra/cmake/FindJsonCpp.cmake
@@ -21,7 +21,7 @@ EXEC_PROGRAM(${CMAKE_CXX_COMPILER}
             OUTPUT_STRIP_TRAILING_WHITESPACE
             )
 
-SET(JSONCPP_NAMES ${JSONCPP_NAMES} libjson_linux-gcc-${_gcc_COMPILER_VERSION}_libmt.so)
+SET(JSONCPP_NAMES ${JSONCPP_NAMES} libjson.so)
 FIND_LIBRARY(JSONCPP_LIBRARY
   NAMES ${JSONCPP_NAMES}
   PATHS /usr/lib /usr/local/lib
